DOC: consolidated coding guide and added naming conventions table (#27213)
matplotlib/matplotlib: doc/devel/contribute.rst

  import numpy as np
  import numpy.ma as ma
  import matplotlib as mpl
  import matplotlib.pyplot as plt
  import matplotlib.cbook as cbook
  import matplotlib.patches as mpatches

In general, Matplotlib modules should **not** import `.rcParams` using ``from
matplotlib import rcParams``, but rather access it as ``mpl.rcParams``.  This
is because some modules are imported very early, before the `.rcParams`
singleton is constructed.

Variable names
--------------

When feasible, please use our internal variable naming convention for objects
of a given class and objects of any child class:

+---------------+------------------------------------+
| variable name |             base class             |
+===============+====================================+
| ``fig``       | `~matplotlib.figure.FigureBase`    |
+---------------+------------------------------------+
| ``ax``        | `~matplotlib.axes.Axes`            |
+---------------+------------------------------------+
| ``trans``     | `~matplotlib.transforms.Transform` |
+---------------+------------------------------------+

Generally, denote more than instance of the same class by adding numbers to

> the variable names, i.e. ``trans1``, ``trans2``.

Reviewer: Can we please suggest naming that is meaningful rather than just numbers?

Reviewer: anything but numbers!  `trans_<from>_to_<target>` ? All of the `trans<name>` attributes are implicitly to screen so (I assume) we dropped the target.  In function bodies where transformations may be targeting other coordinate systems I think we should include both.

| `~matplotlib.figure.FigureBase`    | ``fig``       |                                          |
+------------------------------------+---------------+------------------------------------------+
| `~matplotlib.axes.Axes`            | ``ax``        |                                          |
+------------------------------------+---------------+------------------------------------------+
| `~matplotlib.transforms.Transform` | ``trans``     | ``trans_<source>_<target>``              |
+                                    +               +                                          +
|                                    |               | ``trans_<source>`` when target is screen |
+------------------------------------+---------------+------------------------------------------+

Generally, denote more than one instance of the same class by adding suffixes to
the variable names. If a format isn't specified in the table, use numbers or
letters as appropriate.

.. _type-hints:

Type hints
----------

If you add new public API or change public API, update or add the
corresponding `mypy <https://mypy.readthedocs.io/en/latest/>`_ type hints.

> Most often this is found in the corresponding ``.pyi`` file for the edited
> ``.py`` file, for example ``colors.py`` and ``colors.pyi``. Changes in ``pyplot.py``
> are type hinted inline.

Reviewer 2: ```suggestion
We generally use `stub files
<https://typing.readthedocs.io/en/latest/source/stubs.html#type-stubs>`_
(``*.pyi``) to store the type information; for example ``colors.pyi`` contains
the type information for ``colors.py``. A notable exception is ``pyplot.py``,
which is type hinted inline.
```

Type hints are checked by the mypy :ref:`pre-commit hook <pre-commit-hooks>`
and can often be verified using ``tools\stubtest.py`` and occasionally may
require the use of ``tools\check_typehints.py``.

.. _new-changed-api:

API changes and new features
----------------------------

API consistency and stability are of great value. Therefore, API changes
(e.g. signature changes, behavior changes, removals) will only be conducted
if the added benefit is worth the user effort for adapting.

Because we are a visualization library our primary output is the final

> visualization the user sees.  Thus it is our policy that the appearance of the
> figure is part of the API and any changes, either semantic or esthetic, will be

Reviewer 2: slightly simpler:
```suggestion
visualization the user sees.  We thus regard the appearance of the
figure as part of the API and any changes, either semantic or esthetic, will be
```
